rfcs/2022-02-24-11552-dd-agent-style-secret-management.md

```diff
@@ -0,0 +1,129 @@
+
+## Scope
+
+### In scope
+
+- User shall be able to user the same kind of executable to load/decrypt secrets for Vector.
+- This new feature will have a deterministic behaviour when used in conjunction with templates.
+- Situations like topology reload and load/decryption failure will be accounted for.
+
+### Out of scope
+
+- Integration with other secret or distribution configuration like vault, however this RFC will account for that kind of
+  future extension
+
+## Pain
+
+- As of today secret like authentication tokens and passwords should be provided inside the topology configuration in
+  plain text and that may not be acceptable in some circumstances.
+- Decoupling secret management and key rotation from configuration management.
+
+## Proposal
+
+### User Experience
+
+- Use the same kind of API between Vector and a user-provided executable as the [one between the Agent and the secret
+  retrieving executable][dd-secret-backend-exec-api].
+- A set of top level options like the ones the Datadog Agent [exposes][dd-agent-secret-knobs].
+- A convenient syntax for config option to indicate Vector that a secret should be retrieved for this option (Subject to
+  be changed but the Datadog Agent uses `ENC[secret_key]`)
+
+New top level options to be added:
+
+```toml
+[secret_backend]
+path = "/path/to/the/command"
+argument = "--config foo=bar"
+timeout = 5
+```
+
+Overall the behaviour for corner cases should follow what's in place for environmment variable interpolation as this is
+a very close feature.
+
+### Implementation
+
+A secret backend that would lie in `./src/config/secret.rs` and would call the user provided executable for secrets,
+cache secrets to avoid calling the backend for the same key multiple time. It would read the configuration file once to
+get its config before further processing, ideally env var interpolating should be supported (this should not be a
+problem). In `./src/config/builder.rs`, `load_builder_from_paths` will still be returning a complete configuration with
+placeholders replaced by secrets.
+
+The `ConfigBuilder` struct will get a new `secret_backend` field (type to something like `SecretBackend`), this means
+that `load_builder_from_paths` will then assert if this field is present before returning to the caller, and if it, the
+config will be reloaded with this `SecretBackend` passed to downstream callee and hook the secret interpolation around
+the same point as [the environment variable interpolation][env-var-hook] it would then query this `SecretBackend` for
+each `ENC[secret_key]`.
+
+The implementation should ease future extension and split the internal API queried by the interpolation logic and the
+secret provider that may see other implementation like: `executable` (the one documented in this RFC), `vault`,
+`k8s-config-map`, `aws secretsmanager`, etc.
+
+## Rationale
+
+- Some users just can't put sensitive information inside their configuration.
```

Comment:
```suggestion
- Some users just can't put sensitive information inside their configuration.
- Using environment variable violates security requirements for sensitive environments since environment variables can be leaked by an attacker.
```